File: utils/jwt.go
package utils

import (
	"errors"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

type JWTClaims struct {
	DoctorID string `json:"sub"`
	Email    string `json:"email"`
	jwt.RegisteredClaims
}

func GenerateToken(doctorID, email, secret string, duration time.Duration) (string, error) {
	claims := JWTClaims{
		DoctorID: doctorID,
		Email:    email,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(duration)),
			IssuedAt:  jwt.NewNumericDate(time.Now()),
			Subject:   doctorID,
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	signedToken, err := token.SignedString([]byte(secret))
	if err != nil {
		return "", fmt.Errorf("failed to sign token: %w", err)
	}

	return signedToken, nil
}

func ValidateToken(tokenString, secret string) (*JWTClaims, error) {
	token, err := jwt.ParseWithClaims(tokenString, &JWTClaims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(secret), nil
	})

	if err != nil {
		return nil, err
	}

	claims, ok := token.Claims.(*JWTClaims)
	if !ok || !token.Valid {
		return nil, errors.New("invalid or expired token")
	}

	return claims, nil
}

File: src/utils/jwt.ts
import jwt, { JwtPayload } from 'jsonwebtoken';

export interface JWTClaims extends JwtPayload {
  doctorId: string;
  email: string;
}

export function generateToken(doctorId: string, email: string, secret: string, expiresIn: string = '24h'): string {
  return jwt.sign(
    { sub: doctorId, doctorId, email },
    secret,
    { expiresIn: expiresIn as any }
  );
}

export function validateToken(tokenString: string, secret: string): JWTClaims {
  try {
    const decoded = jwt.verify(tokenString, secret) as any;
    return {
      doctorId: decoded.sub || decoded.doctorId,
      email: decoded.email,
      ...decoded
    };
  } catch (err) {
    throw new Error('Invalid or expired token');
  }
}
